### PR TITLE
Add TCP out of order or duplicate segments sampler via BPF

### PR DESCRIPTION
**Pull request by rittme**
Problem

Add two new TCP probes:
* A probe to measure duplicate TCP segments. This allow us to detect ACK are not being received.
* A probe to measure out of order TCP segments. Out of order segments causes retransmissions and can affect latency.

Solution

Added two BPF kernel probes to count du and RTO events.

Result

Two new metrics introduced to TCP sampler:
* tcp/receive/duplicate
* tcp/receive/out_of_order

### CODE REVIEW

File: src/samplers/tcp/stat.rs [hunk 1]
```diff
@@ -232,6 +238,22 @@ impl TcpStatistic {
...
         };
+        let tcp_duplicate_probe = Probe {
+            name: "tcp_validate_incoming".to_string(),
+            handler: "trace_validate_incoming".to_string(),
+            probe_type: ProbeType::Kernel,
+            probe_location: ProbeLocation::Entry,
+            binary_path: None,
+            sub_system: None,
+        };
+        let tcp_ooo_probe = Probe {
```
Comment (WUMUXIAN) [inline src/samplers/tcp/stat.rs:249]: this probe is identical as the above one. We can just introduce a probe called

tcp_validate_incoming_probe and attach it to both telemetries

Self::Duplicate => vec![tcp_validate_incoming_probe],
Self::OutOfOrder => vec![tcp_validate_incoming_probe],

Comment (rittme) [reply src/samplers/tcp/stat.rs:249]: Cool, I was wondering if I could do that :)

File: src/samplers/tcp/bpf.c [hunk 2]
```diff
@@ -269,3 +272,31 @@ int trace_rto(struct pt_regs *ctx, struct sock *sk)
...
+    if (sk == NULL || skb == NULL)
+        return 0;
+
+    struct tcp_sock *tp = tcp_sk(sk);
+    u32 seq = TCP_SKB_CB(skb)->seq;
+    u32 end_seq = TCP_SKB_CB(skb)->end_seq;
+
+    // Segment sequence before the expected one
+    // which means this was a duplicated segment
+    if ((end_seq - tp->rcv_wup) < 0) {
```
Comment (WUMUXIAN) [inline src/samplers/tcp/bpf.c:287]: I feel the starting sequence number TCP_SKB_CB(skb)->seq and the expected sequence number is tp->rcv_nxt is more straightforward to compare

if TCP_SKB_CB(skb)->seq is smaller than tp->rcv_nxt, it's confirmed a duplicate.

Comment (WUMUXIAN) [reply src/samplers/tcp/bpf.c:287]: 'rcv_wup' records the rcv_nxt on last window update sent, comparing it to the end_seq of this segment does not seem correct to me.

Comment (rittme) [reply src/samplers/tcp/bpf.c:287]: I'm basing these on this function that is used to validate the segment sequence:
https://elixir.bootlin.com/linux/v4.2/source/net/ipv4/tcp_input.c#L3902

From the comments it states:
```
Also, controls (RST is main one) are accepted using RCV.WUP instead 
of RCV.NXT. Peer still did not advance his SND.UNA when we 
delayed ACK, so that hisSND.UNA<=ourRCV.WUP.
```
So to validate it seems better if we use RCV.WUP. But maybe for probing duplicates we don't care about this and should use RCV.NXT anyways?

Comment (WUMUXIAN) [reply src/samplers/tcp/bpf.c:287]: I think what 

```
!before(end_seq, tp->rcv_wup) &&
 !after(seq, tp->rcv_nxt + tcp_receive_window(tp))
```
checks eventually is that whether the sender's segment [seq, end_seq] has any overlays with the expected window at all. 
The first check finds out if the end_seq is even before the previous rcv_nxt sent, the second check finds out if the start seq is even after the largest possible seq (which is rcv_nxt + window size).

For us we simply wants to check if the segment is a duplicate, and checking end_seq - tp->rcv_wup can't help, for example, when (end_seq - tp->rcv_wup) <0, we can confirm that it's a duplicate, but when (end_seq - tp->rcv_wup) > 0, it can still be a duplicate, as long as seq < tp->rcv_nxt. 

So basically to check duplicates, we should be looking seq and tp->rct_nxt.  What do you think?

File: src/samplers/tcp/bpf.c [hunk 3]
```diff
@@ -269,3 +272,31 @@ int trace_rto(struct pt_regs *ctx, struct sock *sk)
...
+        // Increment duplicate counter
+        int loc = 0;
+        add_value(duplicate.lookup(&loc), 1);
+    }
+    
+    // Segment sequence after the expected receive window
+    // which means this segment was received out of order
+    u32 window_end = tp->rcv_nxt + tcp_receive_window(tp);
+    if ((window_end - seq) < 0) {
```
Comment (WUMUXIAN) [inline src/samplers/tcp/bpf.c:296]: Similarly here, I think we can simply check whether if seq > tp->rcv_nxt here to see if the segment arrived out of order. 
window_end < seq only tell that the current segment is totally out of the window, but when there's an overlay, it can't capture.

File: src/samplers/tcp/bpf.c (hunk 2 above)
Comment (rittme) [reply src/samplers/tcp/bpf.c]: Yes, this makes sense. I made the changes to use RCV.NXT instead.

**Comment by brayniac** [comment]
This LGTM from a Rust perspective. @WUMUXIAN - please give a shipit for the BCC portion when you're happy with that and I'll give it another look before merging.

**Review by WUMUXIAN — APPROVED** [review]
looks good to me, I think we can upgrade from draft status and get it merged.

**Comment by rittme** [comment]
@brayniac feel free to merge and close this PR when possible. Thanks!
